Add support for spherical coordinate system in initial topography function

tiannh7 opened the pull request:
Add spherical coordinate system support to initial topography function for improved geometry compatibility.

tiannh7 committed:
Add support for spherical coordinate system in initial topography function

gassmoeller reviewed (changes requested):
Hi @tiannh7 sorry it took a while to get to this PR, I had to think about the logic a bit. I see what you want to achieve as something worth implementing, however I think the current approach you take is more complicated than necessary.

As far as I understand, currently the geometry model hands over the coordinates to the topography model in whichever coordinate system is most natural (box->cartesian, shell->spherical). Then the `value` function always converts these coordinates into cartesian and queries the function for topography.

With your changes the `value` function also converts the coordinates into Cartesian coordinates, but then potentially converts it back to spherical coordinates. I would like to avoid this double conversion, I imagine the logic to be something like this:
```
... input point in natural coordinates

if (is_box_geometry or (is_spherical_geometry and function_takes_spherical_coordinates))
  just copy the point
else if (is_spherical_geometry and not function_takes_spherical_coordinates)
  convert input point to cartesian coordinates
```

that at least saves us one conversion. Also as far as I understand your new change always uses spherical coordinates for spherical geometries. I would prefer if both cartesian or spherical coordinates are possible, similar to how the gravity function model has an input parameter to switch between the two: https://github.com/geodynamics/aspect/blob/main/source/gravity_model/function.cc. If we leave the default value of that parameter at cartesian, we would also keep backward compatibility, for all parameter files out there that rely on the current functionality.

Also yes, please add a test case and an entry in the changelog.

tiannh7 committed:
Adding test stuff

tiannh7 commented:
Hi @gassmoeller,

Thank you very much for your detailed and thoughtful feedback!

I have indeed thought about the approach you mentioned before, but I must admit that I am still not fully familiar with the overall code structure. Your proposed method seems more elegant and efficient.

However, I noticed that in `gravity_model/function.cc`, the function receives the parameter `position` as:

```cpp
...
gravity_vector (const Point<dim> &position) const {
...
  const Utilities::NaturalCoordinate<dim> point =
      this->get_geometry_model().cartesian_to_other_coordinates(position, coordinate_system);
...
}
```
Is this position based on Cartesian coordinates (x, y, z)? On the other hand, in `initial_topography_model/function.cc`, the function requires the parameter `const Point<dim-1> &surface_point`, which has already been processed into natural coordinates before input. Therefore, I implemented a new approach that also avoids the double coordinate conversion you mentioned.

I have already added a test case for this feature. Please let me know if you find any issues or have further suggestions.

Thanks again for your help—I’ve really learned a lot from you!

gassmoeller reviewed (commented):
Yes, this is even better than what I was thinking of. I left a few comments to improve the error checking, but except for the comments this looks pretty good already. Let me know when you have addressed my comments.

tiannh7 commented:
I have fixed the issues as per your suggestions. Thank you very much for pointing out these coding standards. 
I appreciate your thorough review and valuable comments!

gassmoeller commented:
/rebuild

gassmoeller reviewed (approved):
Looks good, thanks for addressing the comments. This is ready to merge when all testers are done and pass.
